### PR TITLE
Conditional processing for VRP endpoints of the HTTP server.

### LINKED ISSUE #469: Feature Request: HTTP serve validation results with Last-Modified header field
I'm currently running tests with various validators, GETting JSON output and post-processing it. It would be very useful if the JSON output would be served with a Last-Modified header field, indicating the time when the last validation run finished.

(see also https://github.com/cloudflare/cfrpki/issues/83)

### PR DESCRIPTION
**Pull request by partim**
This PR adds adding of Etag headers to responses (with Last-Modfied already added previously) and processes If-None-Match and If-Modified-Since headers in requests to all endpoints of the HTTP server providing a list of VRPs.

Resolves #469.

@reschke, if you have the time, maybe you can have a quick look over the implementation of request handling in `src/http.rs`’s `maybe_not_modified` function to check that we follow good HTTP practice?

### CODE REVIEW

**Review by reschke — COMMENTED** [review]
Hm, I'm still bad at reading Rust code.

Two things you may want to check:

- edge case: splitting conditional header fields on "," can fail if the ETag happens to contain a "," (but you might not care as your etags do not)
- matching if-modified-since needs to handle the case where the ts changed within the same 1sec interval (due to the http-date resolution)

I can try to build this and run some blackbox tests though...

**Comment by partim** [comment]
> edge case: splitting conditional header fields on ","

Hm, I guess there could be a case where a legit etag follows later. But then, if this fails, the full content is served, so this fails to the safe side.

> matching if-modified-since needs to handle the case where the ts changed within the same 1sec interval

Yikes. It’s fairly unlikely to happen as real-world validation runs take longer than a second, but I guess we should handle this anyway. This may be something that breaks test infrastructure in fun ways.

**Comment by partim** [comment]
I think this is now handling those two cases properly¹. Thanks for the review!

___
¹ In the case of the etag parsing, it will accept a correct tag followed by garbage, which I think is fine.

**Comment by reschke** [comment]
> In the case of etag parsing, it will accept a correct tag followed by garbage, which I think is fine.

Absolutely; after all, this is a special purpose server. If a client insists on sending ETag not generated by the server, so be it :-)

File: src/http.rs [hunk 1]
```diff
@@ -1021,55 +1002,80 @@ fn vrps(
...
+/// `None`.
+fn maybe_not_modified(
+    req: &Request<Body>,
+    etag: &str,
+    done: DateTime<Utc>,
+) -> Option<Response<Body>> {
+    // First, check If-None-Match.
+    for value in req.headers().get_all("If-None-Match").iter() {
+        // Skip ill-formatted values. By being lazy here we may falsely
+        // return a full response, so this should fine.
```
Comment (DRiKE) [inline src/http.rs:1044]:
```suggestion
        // return a full response, so this should be fine.
```

File: src/http.rs [hunk 2]
```diff
@@ -1232,6 +1265,156 @@ impl Drop for HttpStream {
...
+/// makes this indistinguishable from reaching the end of a correctly
+/// formatted value. As a consequence, we will 304 a request that has the
+/// right tag followed by garbage.
+struct EtagsIter<'a>(&'a str);
+
+impl<'a> Iterator for EtagsIter<'a> {
+    type Item = &'a str;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        // Skip white space and check with if we are done.
```
Comment (DRiKE) [inline src/http.rs:1284]:
```suggestion
        // Skip white space and check if we are done.
```

File: src/http.rs [hunk 3]
```diff
@@ -1232,6 +1265,156 @@ impl Drop for HttpStream {
...
+        }
+
+        Some(res)
+    }
+}
+
+
+//------------ Parsing and Constructing HTTP Dates ---------------------------
+
+/// Definition of the preferred date format (aka IMF-fixedate).
```
Comment (DRiKE) [inline src/http.rs:1324]:
```suggestion
/// Definition of the preferred date format (aka IMF-fixdate).
```

File: src/http.rs [hunk 4]
```diff
@@ -1232,6 +1265,156 @@ impl Drop for HttpStream {
...
+///
+/// The definition allows for relaxed parsing: It accepts additional white
+/// space and ignored case for textual representations. It does, however,
```
Comment (DRiKE) [inline src/http.rs:1327]:
```suggestion
/// space and ignores case for textual representations. It does, however,
```

**Review by DRiKE — APPROVED** [review]
Small typos only, see suggestions
